Ability to configure split button in config.toolbar property

engineering-this opened the pull request:
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## What changes did you make?

This feature is rather straight forward. I've selected base as `t/1679` so GH displays real diff for this branch.

Closes #2091

mlewand commented on plugins/splitbutton/plugin.js:
```diff
@@ -21,10 +21,19 @@
 		if ( face ) {
+			// Split Button can be defined with simpler definition via strings, needed by #2091
+			if ( typeof face === 'string' ) {
+				face = editor.ui.items[ face ];
```
You assigning an object that you modify few lines below. That means you're modifying referenced icon.

We can not do that, you should clone it first and work on it clone instead.

mlewand commented on tests/plugins/splitbutton/splitbutton.js:
```diff
@@ -3,46 +3,108 @@
+							items: items
+						} );
+
+						editor.ui.add( 'customclick', CKEDITOR.UI_SPLITBUTTON, {
```
Code registered as `customclick` handler is duplicated, extract it to common parts.

mlewand commented on tests/plugins/splitbutton/splitbutton.js:
```diff
@@ -3,46 +3,108 @@
+							face: {
+								icon: 'underline',
+								click: function() {
+									counter.faceClick++;
```
The whole `counter` concept could be simplified by using sinon's stubs. Since you're using them in tests, you could just define stubs in test-scope or in context of `tests` object.

Then you'd want to reset it in `setup/teardown` function to make sure that test can not be affected by other unit tests.

engineering-this commented:
I've pushed to this branch commit form #2098.

mlewand commented on plugins/splitbutton/plugin.js:
```diff
@@ -25,6 +25,14 @@
+				face = CKEDITOR.tools.clone( editor.ui.items[ face ] );
+
+				if ( !face.icon ) {
+					face.icon = face.name || face.command;
```
What is the use case for this? If there is no icon in the original button it's fine to leave it without an icon.

engineering-this replied:
Actually this one can be removed, but next case is required, explanations below.

mlewand commented on plugins/splitbutton/plugin.js:
```diff
@@ -36,6 +44,19 @@
+				item = CKEDITOR.tools.clone( editor.ui.items[ item ] );
+
+				if ( !item.icon ) {
+					item.icon = item.name || item.command;
```
Similar case as before.

engineering-this replied:
That is how button plugin works. 
https://github.com/ckeditor/ckeditor-dev/blob/major/plugins/button/plugin.js#L262-L263

```
var name = this.name || this.command,
	iconPath = null,
	overridePath = this.icon;
iconName = name;
```

If there is no `buttonDefinition.icon` property, then icon will fallback to name or command. But `menu plugins` doesn't do that, so in order for menu item to have an icon we need to find that icon same way as `button plugin` does.

Example opening `~/workspace/ckeditor/ckeditor-dev/samples/index.html` and executing following in console:
```
CKEDITOR.instances.editor.ui.get( 'Bold' );
```
Returned object will not have any icon property. Actually when writing this I'm thinking if maybe we should change how button works, so it always have `icon` property when there is icon on button. WDYT @mlewand?

mlewand replied:
As for assigning the `icon` property during button creation - this sounds reasonable if only we started with such approach. Now there might be 3rd party integrations that rely on this property, and rely on it being a "falsy" value. So such change could break some solutions, we don't want to break things.

This rationale sounds reasonable. In this case it make sense to revert this fallback, thanks for clarifying.

Please just drop a short note, or a link to your comment here.

Also make sure to convert the code link above to a perma link, so something like that: https://github.com/ckeditor/ckeditor-dev/blob/6c0469b530cbc767946379e00b5aedcb01e2ba5a/plugins/button/plugin.js#L262-L263

mlewand commented on plugins/splitbutton/plugin.js:
```diff
@@ -94,12 +115,12 @@
 
-			// First button as default. It might be overwritten by actual default button.
-			if ( !defaultButton && !item[ 'default' ] ) {
+			// If no default button was defined then first button will act as default.
```
This logic could be now simplified. Here, in the loop, you could just leave `if ( definition[ 'default' ] && definition[ 'default' ] === item.name ) {` check over here.

Then the second `if` should go outside the loop, and that would check for conditions:
* whether `defaultButton` was not picked, with the logic above ☝️
* whether `definition[ 'default' ]` was not defined at all

Then assign the first item as a default button (if there was at least one item).

mlewand commented on plugins/splitbutton/plugin.js:
```diff
@@ -94,12 +115,12 @@
 				defaultButton = properties.buttons[ item.id ];
 			} else {
-				if ( item[ 'default' ] ) {
+				if ( definition[ 'default' ] && definition[ 'default' ].toLowerCase() === item.name ) {
```
`definition.default` should be given as a button name, thus it should not be case-sensitive. We expect name `Bold`, `Image` etc.

mlewand commented on plugins/toolbar/plugin.js:
```diff
@@ -535,6 +535,10 @@
 					// Ignore items that are configured to be removed.
 					if ( !removeButtons || CKEDITOR.tools.indexOf( removeButtons, name ) == -1 ) {
+						// Option to define UI element in `config.toolbar` (#2091).
+						if ( item.type && item.type in editor.ui._.handlers ) {
```
This feature of supporting type property is nowhere to be found in the API docs. It needs to be reflected in API docs.

mlewand commented on tests/plugins/splitbutton/manual/splitbutton.md:
```diff
@@ -1,5 +1,5 @@
 @bender-ui: collapsed
-@bender-tags: feature, 4.10.0, 1679
+@bender-tags: feature, 4.11.0, 1679
```
Technically this change is out of scope of this issue. It would be nice if you could extract this commit from history and move it to a proper feature branch - but it's not a big deal if it will remain here in case it's tough to do so.

engineering-this commented:
Rebased onto updated #1679 branch. See [comment](https://github.com/ckeditor/ckeditor-dev/pull/1766#issuecomment-405903831).
Also dropped commits updating version number, as it is now corrected in feature branch.

engineering-this commented:
Rebased, as parent branch was rebased also.
